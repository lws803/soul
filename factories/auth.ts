import { Factory } from 'fishery';

import { RefreshToken } from 'src/auth/entities/refresh-token.entity';
import { JWTRefreshPayload } from 'src/auth/entities/jwt-refresh-payload.entity';
import { JWTPayload } from 'src/auth/entities/jwt-payload.entity';
import { TokenType } from 'src/auth/enums/token-type.enum';

import * as factories from './index';

export const refreshToken = Factory.define<RefreshToken>(() => ({
  id: 1,
  user: factories.user.build(),
  isRevoked: false,
  createdAt: new Date('1995-12-17T03:24:00'),
  updatedAt: new Date('1995-12-18T03:24:00'),
  expires: new Date('1995-12-19T03:24:00'),
  platformUser: factories.platformUser.build(),
}));

export const jwtRefreshPayload = Factory.define<JWTRefreshPayload>(() => ({
  tokenId: refreshToken.build().id,
  userId: factories.user.build().id,
  tokenType: TokenType.Refresh,
}));

export const jwtPayload = Factory.define<JWTPayload>(() => {
  const oneUser = factories.user.build();
  return {
    userId: oneUser.id,
    username: oneUser.username,
    tokenType: TokenType.Access,
  };
});

export const jwtPayloadWithPlatform = Factory.define<JWTPayload>(() => {
  const oneUser = factories.user.build();
  const onePlatformUser = factories.platformUser.build({ user: oneUser });
  return {
    userId: oneUser.id,
    username: oneUser.username,
    tokenType: TokenType.Access,
    platformId: onePlatformUser.platform.id,
    roles: onePlatformUser.roles,
  };
});

export const jwtRefreshPayloadWithPlatform = Factory.define<JWTRefreshPayload>(
  () => {
    const oneUser = factories.user.build();
    const onePlatformUser = factories.platformUser.build({ user: oneUser });
    return {
      tokenId: refreshToken.build().id,
      userId: oneUser.id,
      tokenType: TokenType.Refresh,
      platformId: onePlatformUser.platform.id,
      roles: onePlatformUser.roles,
    };
  },
);

export const requestUserObject = Factory.define<{
  tokenType: TokenType;
  username: string;
  userId: number;
}>(() => ({
  tokenType: TokenType.Access,
  username: 'TEST_USER',
  userId: 1,
}));
