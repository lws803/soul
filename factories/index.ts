export * from './users';
export * from './auth';
export * from './platform';
export * from './user-connections';
