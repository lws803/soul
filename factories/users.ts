import { Factory } from 'fishery';

import { CreateUserDto, UpdateUserDto } from 'src/users/serializers/api.dto';
import { User } from 'src/users/entities/user.entity';

export const user = Factory.define<User>(({ sequence }) => {
  // Rewinds the sequence as we do not want to have "sequences" bleed into subsequent builds
  user.rewindSequence();
  return {
    id: sequence,
    username: `test-user-${sequence}`,
    userHandle: `test-user-${sequence}#${sequence}`,
    email: `TEST_USER_${sequence}@EMAIL.COM`,
    isActive: true,
    createdAt: new Date('1995-12-17T03:24:00'),
    updatedAt: new Date('1995-12-18T03:24:00'),
    hashedPassword: 'TEST_HASHED_PASSWORD',
    displayName: 'DISPLAY_NAME',
    bio: 'BIO',
  };
});

export const createUserDto = Factory.define<CreateUserDto>(() => ({
  email: 'TEST_USER@EMAIL.COM',
  username: 'test-user',
  password: 'TEST_PASSWORD',
  bio: null,
  displayName: null,
}));

export const updateUserDto = Factory.define<UpdateUserDto>(() => ({
  email: 'UPDATED_EMAIL@EMAIL.COM',
  username: 'updated-user',
  displayName: 'UPDATED_DISPLAY_NAME',
  bio: 'UPDATED_BIO',
}));
