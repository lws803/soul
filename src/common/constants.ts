export const DEFAULT_NUM_ITEMS_PER_PAGE = 10;
