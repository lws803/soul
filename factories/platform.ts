import { Factory } from 'fishery';

import * as factories from 'factories';
import { Platform } from 'src/platforms/entities/platform.entity';
import { PlatformUser } from 'src/platforms/entities/platform-user.entity';
import {
  UpdatePlatformDto,
  CreatePlatformDto,
} from 'src/platforms/dto/api.dto';
import { UserRole } from 'src/roles/role.enum';
import { PlatformCategory } from 'src/platforms/entities/platform-category.entity';

export const onePlatformCategory = Factory.define<PlatformCategory>(() => ({
  id: 1,
  name: 'CATEGORY',
}));

export const onePlatform = Factory.define<Platform>(() => ({
  id: 1,
  name: 'TEST_PLATFORM',
  nameHandle: 'test_platform#1',
  createdAt: new Date('1995-12-17T03:24:00'),
  updatedAt: new Date('1995-12-18T03:24:00'),
  userConnections: [],
  isVerified: true,
  redirectUris: ['TEST_REDIRECT_URI'],
  category: onePlatformCategory.build(),
  activityWebhookUri: 'ACTIVITY_WEBHOOK_URI',
}));

export const platformArray = Factory.define<Platform[]>(() => [
  onePlatform.build(),
  onePlatform.build({ id: 2, name: 'TEST_PLATFORM_2' }),
]);

export const onePlatformUser = Factory.define<PlatformUser>(() => ({
  id: 1,
  user: factories.oneUser.build(),
  platform: factories.onePlatform.build(),
  roles: [UserRole.Admin, UserRole.Member],
  createdAt: new Date('1995-12-17T03:24:00'),
  updatedAt: new Date('1995-12-18T03:24:00'),
}));

export const platformUserArray = Factory.define<PlatformUser[]>(() => [
  onePlatformUser.build(),
  onePlatformUser.build({ id: 2 }),
]);

export const updatePlatformDto = Factory.define<UpdatePlatformDto>(() => ({
  name: 'TEST_PLATFORM_UPDATE',
  category: 'CATEGORY_UPDATE',
}));

export const createPlatformDto = Factory.define<CreatePlatformDto>(() => ({
  name: 'TEST_PLATFORM',
  redirectUris: ['TEST_REDIRECT_URI'],
  category: onePlatformCategory.build().name,
}));

type CreatePlatformRequest = {
  name: string;
  redirect_uris: string[];
  category: string;
};

export const createPlatformRequestDto = Factory.define<CreatePlatformRequest>(
  () => ({
    name: 'TEST_PLATFORM',
    redirect_uris: ['TEST_REDIRECT_URI'],
    category: onePlatformCategory.build().name,
  }),
);
