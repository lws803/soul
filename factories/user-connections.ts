import { Factory } from 'fishery';

import { FindOneUserConnectionResponseEntity } from 'src/user-connections/serializers/api-responses.entity';

import { UserConnection } from 'src/user-connections/entities/user-connection.entity';

import { platformEntity } from './platform';
import { userEntity } from './users';

export const userConnectionEntity = Factory.define<UserConnection>(() => ({
  id: 1,
  fromUser: userEntity.build(),
  toUser: userEntity.build({ id: 2, email: 'TEST_USER_2@EMAIL.COM' }),
  platforms: [],
  createdAt: new Date('1995-12-17T03:24:00'),
  updatedAt: new Date('1995-12-18T03:24:00'),
  mutualConnection: null,
}));

export const oneUserConnectionResponseEntity =
  Factory.define<FindOneUserConnectionResponseEntity>(() => ({
    id: 1,
    fromUser: userEntity.build(),
    toUser: userEntity.build({ id: 2, email: 'TEST_USER_2@EMAIL.COM' }),
    platforms: [],
    createdAt: new Date('1995-12-17T03:24:00'),
    updatedAt: new Date('1995-12-18T03:24:00'),
    mutualConnection: null,
    isMutual: false,
  }));

export const userConnectionEntityArray = Factory.define<UserConnection[]>(
  () => [
    userConnectionEntity.build({ id: 1 }),
    userConnectionEntity.build({
      id: 2,
      fromUser: userEntity.build({ id: 3 }),
      toUser: userEntity.build({ id: 4 }),
    }),
  ],
);

type CreateUserConnectionRequest = {
  to_user_id: number;
  platform_id?: number;
};

export const createUserConnectionRequest =
  Factory.define<CreateUserConnectionRequest>(() => ({
    to_user_id: userEntity.build({ id: 2 }).id,
  }));

type PostPlatformToUserConnectionRequest = {
  platform_id: number;
};

export const postPlatformToUserConnectionRequest =
  Factory.define<PostPlatformToUserConnectionRequest>(() => ({
    platform_id: platformEntity.build().id,
  }));
