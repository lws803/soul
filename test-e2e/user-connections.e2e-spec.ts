import { HttpStatus, INestApplication } from '@nestjs/common';
import { Connection, Repository } from 'typeorm';
import * as request from 'supertest';

import { User } from 'src/users/entities/user.entity';
import { UserConnection } from 'src/user-connections/entities/user-connection.entity';
import { Platform } from 'src/platforms/entities/platform.entity';
import { PlatformUser } from 'src/platforms/entities/platform-user.entity';
import { PlatformCategory } from 'src/platforms/entities/platform-category.entity';

import * as factories from '../factories';

import createAppFixture from './fixtures/create-app-fixture';
import { createUsersAndLoginFixture } from './fixtures/create-users-and-login-fixture';

describe('UserConnectionsController (e2e)', () => {
  let app: INestApplication;
  let connection: Connection;
  let userRepository: Repository<User>;
  let userConnectionRepository: Repository<UserConnection>;
  let platformRepository: Repository<Platform>;
  let platformUserRepository: Repository<PlatformUser>;
  let platformCategoryRepository: Repository<PlatformCategory>;

  let firstUserAccessToken: string;

  beforeAll(async () => {
    app = await createAppFixture({});
    await app.init();
    app.useLogger(false);

    connection = app.get(Connection);
    userRepository = connection.getRepository(User);
    userConnectionRepository = connection.getRepository(UserConnection);
    platformRepository = connection.getRepository(Platform);
    platformUserRepository = connection.getRepository(PlatformUser);
    platformCategoryRepository = connection.getRepository(PlatformCategory);

    await connection.synchronize(true);

    const [{ accessToken }] = await createUsersAndLoginFixture(app);
    firstUserAccessToken = accessToken;

    await platformCategoryRepository.save(
      factories.onePlatformCategory.build(),
    );
  });

  afterAll(async () => {
    await app.close();
  });

  describe('/user-connections (POST)', () => {
    afterEach(async () => {
      await userConnectionRepository.delete({});
      await platformUserRepository.delete({});
      await platformRepository.delete({});
    });

    it('creates a new connection', async () => {
      return request(app.getHttpServer())
        .post('/user-connections')
        .set('Authorization', `Bearer ${firstUserAccessToken}`)
        .send(factories.createUserConnectionRequest.build())
        .expect(HttpStatus.CREATED)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            created_at: expect.any(String),
            updated_at: expect.any(String),
            id: expect.any(Number),
            from_user: {
              id: 1,
              user_handle: 'test-user#1',
              username: 'test-user',
              bio: null,
              display_name: null,
            },
            to_user: {
              id: 2,
              user_handle: 'test-user-2#2',
              username: 'test-user-2',
              bio: null,
              display_name: null,
            },
            platforms: [],
            is_mutual: false,
          }),
        );
    });

    it('creates a new connection with platform', async () => {
      const platform = factories.platform.build();
      await platformRepository.save(platform);
      await platformUserRepository.save(
        factories.platformUser.build({ platform }),
      );

      return request(app.getHttpServer())
        .post('/user-connections')
        .set('Authorization', `Bearer ${firstUserAccessToken}`)
        .send(factories.createUserConnectionRequest.build({ platform_id: 1 }))
        .expect(HttpStatus.CREATED)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            created_at: expect.any(String),
            updated_at: expect.any(String),
            id: expect.any(Number),
            from_user: {
              id: 1,
              user_handle: 'test-user#1',
              username: 'test-user',
              bio: null,
              display_name: null,
            },
            to_user: {
              id: 2,
              user_handle: 'test-user-2#2',
              username: 'test-user-2',
              bio: null,
              display_name: null,
            },
            platforms: [
              {
                created_at: expect.any(String),
                updated_at: expect.any(String),
                id: expect.any(Number),
                is_verified: true,
                name: platform.name,
                name_handle: platform.nameHandle,
                category: {
                  id: 1,
                  name: 'CATEGORY',
                },
              },
            ],
            is_mutual: false,
          }),
        );
    });

    it('creates a new connection with opposite becomes mutual', async () => {
      const secondUserConnection = factories.oneUserConnection.build({
        id: 2,
        fromUser: factories.user.build({
          id: 2,
          email: 'TEST_USER_2@EMAIL.COM',
        }),
        toUser: factories.user.build(),
      });
      await userConnectionRepository.save(secondUserConnection);
      await userConnectionRepository.update(
        { id: 1 },
        { mutualConnection: secondUserConnection },
      );

      return request(app.getHttpServer())
        .post('/user-connections')
        .set('Authorization', `Bearer ${firstUserAccessToken}`)
        .send(factories.createUserConnectionRequest.build())
        .expect(HttpStatus.CREATED)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            created_at: expect.any(String),
            updated_at: expect.any(String),
            id: expect.any(Number),
            from_user: {
              id: 1,
              user_handle: 'test-user#1',
              username: 'test-user',
              bio: null,
              display_name: null,
            },
            to_user: {
              id: 2,
              user_handle: 'test-user-2#2',
              username: 'test-user-2',
              bio: null,
              display_name: null,
            },
            platforms: [],
            is_mutual: true,
          }),
        );
    });
  });

  describe('/user-connections/by-users (GET)', () => {
    beforeAll(async () => {
      const firstUserConnection = factories.oneUserConnection.build();
      const secondUserConnection = factories.oneUserConnection.build({
        id: 2,
        fromUser: factories.user.build({
          id: 2,
          email: 'TEST_USER_2@EMAIL.COM',
        }),
        toUser: factories.user.build(),
      });
      await userConnectionRepository.save([
        firstUserConnection,
        secondUserConnection,
      ]);
      await userConnectionRepository.save([
        { ...firstUserConnection, mutualConnection: secondUserConnection },
        { ...secondUserConnection, mutualConnection: firstUserConnection },
      ]);
    });
    afterAll(async () => {
      await userConnectionRepository.delete({});
    });

    it('fetches user connection by users', async () => {
      return request(app.getHttpServer())
        .get('/user-connections/by-users?from_user_id=1&to_user_id=2')
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            created_at: expect.any(String),
            updated_at: expect.any(String),
            from_user: {
              id: 1,
              user_handle: 'test-user#1',
              username: 'test-user',
              bio: null,
              display_name: null,
            },
            id: 1,
            platforms: [],
            to_user: {
              id: 2,
              user_handle: 'test-user-2#2',
              username: 'test-user-2',
              bio: null,
              display_name: null,
            },
            is_mutual: true,
          }),
        );
    });

    it('returns not found', async () => {
      await userRepository.save([
        factories.user.build({
          id: 999,
          email: 'TEST_USER_999@EMAIL.COM',
          userHandle: 'test-user_999#999',
          username: 'test-user-999',
        }),
        factories.user.build({
          id: 998,
          email: 'TEST_USER_998@EMAIL.COM',
          userHandle: 'test-user_998#998',
          username: 'test-user-998',
        }),
      ]);
      return request(app.getHttpServer())
        .get('/user-connections/by-users?from_user_id=999&to_user_id=998')
        .expect(404)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            error: 'USER_CONNECTION_NOT_FOUND',
            message: 'User connection not found',
          }),
        );
    });
  });

  describe('/user-connections/:id (GET)', () => {
    beforeAll(async () => {
      const firstUserConnection = factories.oneUserConnection.build();
      const secondUserConnection = factories.oneUserConnection.build({
        id: 2,
        fromUser: factories.user.build({
          id: 2,
          email: 'TEST_USER_2@EMAIL.COM',
        }),
        toUser: factories.user.build(),
      });
      await userConnectionRepository.save([
        firstUserConnection,
        secondUserConnection,
      ]);
      await userConnectionRepository.save([
        { ...firstUserConnection, mutualConnection: secondUserConnection },
        { ...secondUserConnection, mutualConnection: firstUserConnection },
      ]);
    });
    afterAll(async () => {
      await userConnectionRepository.delete({});
    });

    it('fetches user connection', async () => {
      return request(app.getHttpServer())
        .get('/user-connections/1')
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            created_at: expect.any(String),
            updated_at: expect.any(String),
            from_user: {
              id: 1,
              user_handle: 'test-user#1',
              username: 'test-user',
              bio: null,
              display_name: null,
            },
            id: 1,
            platforms: [],
            to_user: {
              id: 2,
              user_handle: 'test-user-2#2',
              username: 'test-user-2',
              bio: null,
              display_name: null,
            },
            is_mutual: true,
          }),
        );
    });

    it('returns not found', async () => {
      return request(app.getHttpServer())
        .get('/user-connections/999')
        .expect(HttpStatus.NOT_FOUND)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            error: 'USER_CONNECTION_NOT_FOUND',
            message: 'User connection with id: 999 not found',
          }),
        );
    });
  });

  describe('/user-connections/:id (DELETE)', () => {
    beforeEach(async () => {
      await userConnectionRepository.save(factories.oneUserConnection.build());
    });

    afterEach(async () => {
      await userConnectionRepository.delete({});
    });

    it('deletes user connection', async () => {
      return request(app.getHttpServer())
        .delete('/user-connections/1')
        .set('Authorization', `Bearer ${firstUserAccessToken}`)
        .expect(HttpStatus.OK)
        .expect((res) => expect(res.body).toStrictEqual({}));
    });

    it('deleting user connection removes mutual status', async () => {
      const firstUserConnection = factories.oneUserConnection.build();
      const secondUserConnection = factories.oneUserConnection.build({
        id: 2,
        fromUser: factories.user.build({
          id: 2,
          email: 'TEST_USER_2@EMAIL.COM',
        }),
        toUser: factories.user.build(),
      });
      await userConnectionRepository.save([
        firstUserConnection,
        secondUserConnection,
      ]);
      await userConnectionRepository.save([
        { ...firstUserConnection, mutualConnection: secondUserConnection },
        { ...secondUserConnection, mutualConnection: firstUserConnection },
      ]);

      await request(app.getHttpServer())
        .delete('/user-connections/1')
        .set('Authorization', `Bearer ${firstUserAccessToken}`)
        .expect(HttpStatus.OK)
        .expect((res) => expect(res.body).toStrictEqual({}));

      return request(app.getHttpServer())
        .get('/user-connections/2')
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            created_at: expect.any(String),
            updated_at: expect.any(String),
            from_user: {
              id: 2,
              user_handle: 'test-user-2#2',
              username: 'test-user-2',
              bio: null,
              display_name: null,
            },
            id: 2,
            platforms: [],
            to_user: {
              id: 1,
              user_handle: 'test-user#1',
              username: 'test-user',
              bio: null,
              display_name: null,
            },
            is_mutual: false,
          }),
        );
    });
  });

  describe('/user-connections/:id/platforms (POST)', () => {
    beforeEach(async () => {
      await userConnectionRepository.save(factories.oneUserConnection.build());
    });

    afterEach(async () => {
      await userConnectionRepository.delete({});
      await platformUserRepository.delete({});
      await platformRepository.delete({});
    });

    it('adds a new platform to the existing connection', async () => {
      const platform = factories.platform.build();
      await platformRepository.save(platform);
      await platformUserRepository.save(factories.platformUser.build());

      await request(app.getHttpServer())
        .post('/user-connections/1/platforms')
        .set('Authorization', `Bearer ${firstUserAccessToken}`)
        .send(factories.postPlatformToUserConnectionRequest.build())
        .expect(HttpStatus.CREATED)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            id: 1,
            created_at: expect.any(String),
            updated_at: expect.any(String),
            from_user: {
              id: 1,
              user_handle: 'test-user#1',
              username: 'test-user',
              bio: null,
              display_name: null,
            },
            to_user: {
              id: 2,
              user_handle: 'test-user-2#2',
              username: 'test-user-2',
              bio: null,
              display_name: null,
            },
            platforms: [
              {
                created_at: expect.any(String),
                updated_at: expect.any(String),
                id: 1,
                is_verified: true,
                name: platform.name,
                name_handle: platform.nameHandle,
                category: {
                  id: 1,
                  name: 'CATEGORY',
                },
              },
            ],
          }),
        );
    });
  });

  describe('/user-connections/:id/platforms (DELETE)', () => {
    beforeEach(async () => {
      await userConnectionRepository.save(factories.oneUserConnection.build());
    });

    afterEach(async () => {
      await userConnectionRepository.delete({});
      await platformUserRepository.delete({});
      await platformRepository.delete({});
    });

    it('deletes a platform from existing connection', async () => {
      await platformRepository.save(factories.platform.build());
      await platformUserRepository.save(factories.platformUser.build());

      return request(app.getHttpServer())
        .delete('/user-connections/1/platforms/1')
        .set('Authorization', `Bearer ${firstUserAccessToken}`)
        .expect(HttpStatus.OK)
        .expect((res) => expect(res.body).toStrictEqual({}));
    });
  });

  describe('/user-connections/my-connections (GET)', () => {
    afterEach(async () => {
      await userConnectionRepository.delete({});
    });

    it('fetches my mutual connections', async () => {
      const firstUserConnection = factories.oneUserConnection.build();
      const secondUserConnection = factories.oneUserConnection.build({
        id: 2,
        fromUser: factories.user.build({
          id: 2,
          email: 'TEST_USER_2@EMAIL.COM',
        }),
        toUser: factories.user.build(),
      });
      await userConnectionRepository.save([
        firstUserConnection,
        secondUserConnection,
      ]);
      await userConnectionRepository.save([
        { ...firstUserConnection, mutualConnection: secondUserConnection },
        { ...secondUserConnection, mutualConnection: firstUserConnection },
      ]);

      return request(app.getHttpServer())
        .get('/user-connections/my-connections?connection_type=mutual')
        .set('Authorization', `Bearer ${firstUserAccessToken}`)
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            total_count: 1,
            user_connections: [
              {
                created_at: expect.any(String),
                updated_at: expect.any(String),
                from_user: {
                  id: 1,
                  user_handle: 'test-user#1',
                  username: 'test-user',
                  bio: null,
                  display_name: null,
                },
                id: expect.any(Number),
                platforms: [],
                to_user: {
                  id: 2,
                  user_handle: 'test-user-2#2',
                  username: 'test-user-2',
                  bio: null,
                  display_name: null,
                },
              },
            ],
          }),
        );
    });

    it('fetches my follower connections', async () => {
      const secondUserConnection = factories.oneUserConnection.build({
        id: 2,
        fromUser: factories.user.build({
          id: 2,
          email: 'TEST_USER_2@EMAIL.COM',
        }),
        toUser: factories.user.build(),
      });
      await userConnectionRepository.save(secondUserConnection);

      return request(app.getHttpServer())
        .get('/user-connections/my-connections?connection_type=follower')
        .set('Authorization', `Bearer ${firstUserAccessToken}`)
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            total_count: 1,
            user_connections: [
              {
                created_at: expect.any(String),
                updated_at: expect.any(String),
                to_user: {
                  id: 1,
                  user_handle: 'test-user#1',
                  username: 'test-user',
                  bio: null,
                  display_name: null,
                },
                id: expect.any(Number),
                platforms: [],
                from_user: {
                  id: 2,
                  user_handle: 'test-user-2#2',
                  username: 'test-user-2',
                  bio: null,
                  display_name: null,
                },
              },
            ],
          }),
        );
    });

    it('fetches my follow connections', async () => {
      const firstUserConnection = factories.oneUserConnection.build();
      await userConnectionRepository.save(firstUserConnection);

      return request(app.getHttpServer())
        .get('/user-connections/my-connections?connection_type=following')
        .set('Authorization', `Bearer ${firstUserAccessToken}`)
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            total_count: 1,
            user_connections: [
              {
                created_at: expect.any(String),
                updated_at: expect.any(String),
                from_user: {
                  id: 1,
                  user_handle: 'test-user#1',
                  username: 'test-user',
                  bio: null,
                  display_name: null,
                },
                id: expect.any(Number),
                platforms: [],
                to_user: {
                  id: 2,
                  user_handle: 'test-user-2#2',
                  username: 'test-user-2',
                  bio: null,
                  display_name: null,
                },
              },
            ],
          }),
        );
    });

    it('fetches my follow connections for a platform', async () => {
      const thirdUser = factories.user.build({
        id: 3,
        username: 'test-user_3',
        userHandle: 'test-user_3#3',
        email: 'TEST_USER_3@EMAIL.COM',
      });
      const onePlatform = factories.platform.build();
      await userRepository.save(thirdUser);
      await platformRepository.save(onePlatform);
      const firstUserConnection = factories.oneUserConnection.build({
        platforms: [],
      });
      const secondUserConnection = factories.oneUserConnection.build({
        id: 2,
        toUser: thirdUser,
        platforms: factories.platform.buildList(1),
      });
      await userConnectionRepository.save([
        firstUserConnection,
        secondUserConnection,
      ]);

      return request(app.getHttpServer())
        .get(
          '/user-connections/my-connections?connection_type=following&platform_id=1',
        )
        .set('Authorization', `Bearer ${firstUserAccessToken}`)
        .expect(HttpStatus.OK)
        .expect((res) =>
          expect(res.body).toStrictEqual({
            total_count: 1,
            user_connections: [
              {
                id: expect.any(Number),
                created_at: expect.any(String),
                updated_at: expect.any(String),
                from_user: {
                  id: 1,
                  user_handle: 'test-user#1',
                  username: 'test-user',
                  bio: null,
                  display_name: null,
                },
                platforms: [
                  {
                    created_at: expect.any(String),
                    updated_at: expect.any(String),
                    id: 1,
                    is_verified: true,
                    name: onePlatform.name,
                    name_handle: onePlatform.nameHandle,
                  },
                ],
                to_user: {
                  id: 3,
                  user_handle: 'test-user_3#3',
                  username: 'test-user_3',
                  bio: null,
                  display_name: null,
                },
              },
            ],
          }),
        );
    });
  });
});
