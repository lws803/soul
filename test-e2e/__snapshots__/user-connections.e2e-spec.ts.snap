// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserConnectionsController (e2e) /user-connections (POST) creates a new connection 1`] = `
Object {
  "created_at": Any<String>,
  "from_user": Object {
    "id": 1,
    "user_handle": "test_user#1",
    "username": "TEST_USER",
  },
  "id": Any<Number>,
  "is_mutual": false,
  "platforms": Array [],
  "to_user": Object {
    "id": 2,
    "user_handle": "test_user_2#2",
    "username": "TEST_USER_2",
  },
  "updated_at": Any<String>,
}
`;

exports[`UserConnectionsController (e2e) /user-connections (POST) creates a new connection with opposite becomes mutual 1`] = `
Object {
  "created_at": Any<String>,
  "from_user": Object {
    "id": 1,
    "user_handle": "test_user#1",
    "username": "TEST_USER",
  },
  "id": Any<Number>,
  "is_mutual": true,
  "platforms": Array [],
  "to_user": Object {
    "id": 2,
    "user_handle": "test_user_2#2",
    "username": "TEST_USER_2",
  },
  "updated_at": Any<String>,
}
`;

exports[`UserConnectionsController (e2e) /user-connections (POST) creates a new connection with platform 1`] = `
Object {
  "created_at": Any<String>,
  "from_user": Object {
    "id": 1,
    "user_handle": "test_user#1",
    "username": "TEST_USER",
  },
  "id": Any<Number>,
  "is_mutual": false,
  "platforms": Array [
    Object {
      "category": Object {
        "id": 1,
        "name": "CATEGORY",
      },
      "created_at": Any<String>,
      "id": Any<Number>,
      "is_verified": true,
      "name": "TEST_PLATFORM_1",
      "name_handle": "test_platform_1#1",
      "updated_at": Any<String>,
    },
  ],
  "to_user": Object {
    "id": 2,
    "user_handle": "test_user_2#2",
    "username": "TEST_USER_2",
  },
  "updated_at": Any<String>,
}
`;

exports[`UserConnectionsController (e2e) /user-connections/:id (DELETE) deleting user connection removes mutual status 1`] = `
Object {
  "created_at": Any<String>,
  "from_user": Object {
    "id": 2,
    "user_handle": "test_user_2#2",
    "username": "TEST_USER_2",
  },
  "id": 2,
  "is_mutual": false,
  "platforms": Array [],
  "to_user": Object {
    "id": 1,
    "user_handle": "test_user#1",
    "username": "TEST_USER",
  },
  "updated_at": Any<String>,
}
`;

exports[`UserConnectionsController (e2e) /user-connections/:id (GET) fetches user connection 1`] = `
Object {
  "created_at": Any<String>,
  "from_user": Object {
    "id": 1,
    "user_handle": "test_user#1",
    "username": "TEST_USER",
  },
  "id": 1,
  "is_mutual": true,
  "platforms": Array [],
  "to_user": Object {
    "id": 2,
    "user_handle": "test_user_2#2",
    "username": "TEST_USER_2",
  },
  "updated_at": Any<String>,
}
`;

exports[`UserConnectionsController (e2e) /user-connections/:id/platforms (POST) adds a new platform to the existing connection 1`] = `
Object {
  "created_at": Any<String>,
  "from_user": Object {
    "id": 1,
    "user_handle": "test_user#1",
    "username": "TEST_USER",
  },
  "id": 1,
  "platforms": Array [
    Object {
      "category": Object {
        "id": 1,
        "name": "CATEGORY",
      },
      "created_at": Any<String>,
      "id": 1,
      "is_verified": true,
      "name": "TEST_PLATFORM_1",
      "name_handle": "test_platform_1#1",
      "updated_at": Any<String>,
    },
  ],
  "to_user": Object {
    "id": 2,
    "user_handle": "test_user_2#2",
    "username": "TEST_USER_2",
  },
  "updated_at": Any<String>,
}
`;

exports[`UserConnectionsController (e2e) /user-connections/by-users (GET) fetches user connection by users 1`] = `
Object {
  "created_at": Any<String>,
  "from_user": Object {
    "id": 1,
    "user_handle": "test_user#1",
    "username": "TEST_USER",
  },
  "id": 1,
  "is_mutual": true,
  "platforms": Array [],
  "to_user": Object {
    "id": 2,
    "user_handle": "test_user_2#2",
    "username": "TEST_USER_2",
  },
  "updated_at": Any<String>,
}
`;

exports[`UserConnectionsController (e2e) /user-connections/my-connections (GET) fetches my follow connections 1`] = `
Object {
  "total_count": 1,
  "user_connections": Array [
    Object {
      "created_at": Any<String>,
      "from_user": Object {
        "id": 1,
        "user_handle": "test_user#1",
        "username": "TEST_USER",
      },
      "id": Any<Number>,
      "platforms": Array [],
      "to_user": Object {
        "id": 2,
        "user_handle": "test_user_2#2",
        "username": "TEST_USER_2",
      },
      "updated_at": Any<String>,
    },
  ],
}
`;

exports[`UserConnectionsController (e2e) /user-connections/my-connections (GET) fetches my follow connections for a platform 1`] = `
Object {
  "total_count": 1,
  "user_connections": Array [
    Object {
      "created_at": Any<String>,
      "from_user": Object {
        "id": 1,
        "user_handle": "test_user#1",
        "username": "TEST_USER",
      },
      "id": Any<Number>,
      "platforms": Array [
        Object {
          "created_at": Any<String>,
          "id": 1,
          "is_verified": true,
          "name": "TEST_PLATFORM_1",
          "name_handle": "test_platform_1#1",
          "updated_at": Any<String>,
        },
      ],
      "to_user": Object {
        "id": 3,
        "user_handle": "TEST_USER_3#3",
        "username": "TEST_USER_3",
      },
      "updated_at": Any<String>,
    },
  ],
}
`;

exports[`UserConnectionsController (e2e) /user-connections/my-connections (GET) fetches my follower connections 1`] = `
Object {
  "total_count": 1,
  "user_connections": Array [
    Object {
      "created_at": Any<String>,
      "from_user": Object {
        "id": 2,
        "user_handle": "test_user_2#2",
        "username": "TEST_USER_2",
      },
      "id": Any<Number>,
      "platforms": Array [],
      "to_user": Object {
        "id": 1,
        "user_handle": "test_user#1",
        "username": "TEST_USER",
      },
      "updated_at": Any<String>,
    },
  ],
}
`;

exports[`UserConnectionsController (e2e) /user-connections/my-connections (GET) fetches my mutual connections 1`] = `
Object {
  "total_count": 1,
  "user_connections": Array [
    Object {
      "created_at": Any<String>,
      "from_user": Object {
        "id": 1,
        "user_handle": "test_user#1",
        "username": "TEST_USER",
      },
      "id": Any<Number>,
      "platforms": Array [],
      "to_user": Object {
        "id": 2,
        "user_handle": "test_user_2#2",
        "username": "TEST_USER_2",
      },
      "updated_at": Any<String>,
    },
  ],
}
`;
